frame/example-offchain-worker/src/tests.rs

```diff
@@ -71,22 +75,36 @@ impl frame_system::Trait for Test {
 }
 
 type Extrinsic = TestXt<Call<Test>, ()>;
-type SubmitTransaction = frame_system::offchain::TransactionSubmitter<
-	crypto::Public,
-	Test,
-	Extrinsic
->;
-
-impl frame_system::offchain::CreateTransaction<Test, Extrinsic> for Test {
-	type Public = sp_core::sr25519::Public;
-	type Signature = sp_core::sr25519::Signature;
-
-	fn create_transaction<F: frame_system::offchain::Signer<Self::Public, Self::Signature>>(
-		call: <Extrinsic as ExtrinsicsT>::Call,
-		_public: Self::Public,
-		_account: <Test as frame_system::Trait>::AccountId,
-		nonce: <Test as frame_system::Trait>::Index,
-	) -> Option<(<Extrinsic as ExtrinsicsT>::Call, <Extrinsic as ExtrinsicsT>::SignaturePayload)> {
+type AccountId = <<Signature as Verify>::Signer as IdentifyAccount>::AccountId;
+
+impl frame_system::offchain::SigningTypes for Test {
+	type Public = <Signature as Verify>::Signer;
+	type Signature = Signature;
+}
+
+impl<LocalCall> frame_system::offchain::SendTransactionTypes<LocalCall> for Test where
+	Call<Test>: From<LocalCall>,
+{
+	type OverarchingCall = Call<Test>;
+	type Extrinsic = Extrinsic;
+}
+
+pub struct TestAuthId;
+impl frame_system::offchain::AppCrypto<<Signature as Verify>::Signer, Signature> for TestAuthId {
+	type RuntimeAppPublic = crypto::Public;
+	type GenericSignature = sp_core::sr25519::Signature;
+	type GenericPublic = sp_core::sr25519::Public;
+}
+
+impl<LocalCall> frame_system::offchain::CreateSignedTransaction<LocalCall> for Test where
+	Call<Test>: From<LocalCall>,
+{
+	fn create_transaction<C: frame_system::offchain::AppCrypto<Self::Public, Self::Signature>>(
+		call: Call<Test>,
+		_public: <Signature as Verify>::Signer,
+		_account: AccountId,
+		nonce: u64,
+	) -> Option<(Call<Test>, <Extrinsic as ExtrinsicT>::SignaturePayload)> {
```

Comment:
So here is what I read: 

- via `SigningTypes ` I define the runtime/chain's key types. This seems to happen once per runtime and it makes sense as we only have one crypto per chain. 

- via `SendTransactionTypes ` I declare the call type. This is once per runtime and it makes sense to stick it to the runtime object, `Test` in this case. 

- I think `TestAuthId` can be skipped and maybe implement this `offchain:: AppCrypto ` for `crypto::AuthorityId` in the `mod::crypto` and get it dealt with over there?

Reply:
I think this is more like a code organization suggestion. I moved `AppCrypto` implementation to `mod::crypto` but notice the clash in `Signature` naming. I would have preferred keeping it outside where the offchain definitions are in close proximity.